journal/types.go

```diff
@@ -0,0 +1,60 @@
+package journal
+
+import "time"
+
+// EventType represents the signature of an event.
+type EventType struct {
+	System string
+	Event  string
+}
+
+// Journal represents an audit trail of system actions.
+//
+// Every entry is tagged with a timestamp, a system name, and an event name.
+// The supplied data can be any type, as long as it is JSON serializable,
+// including structs, map[string]interface{}, or primitive types.
+//
+// For cleanliness and type safety, we recommend to use typed events. See the
+// *Evt struct types in this package for more info.
+type Journal interface {
+	// IsEnabled allows components to check if a given event type is enabled.
+	// All event types are enabled by default, and specific event types can only
+	// be disabled at construction type. Components are advised to check if the
+	// journal event types they record are enabled as soon as possible, and hold
+	// on to the answer all throughout the lifetime of the process.
+	IsEnabled(evtType EventType) bool
```

Comment:
Why not just have `AddEntry` ignore disabled events?

Reply:
Because (a) that would require a map lookup if implemented naively, and (b) by that time, the caller has already constructed the data object.

I think a cleaner, more ergonomic approach would be to replace this with a `Journal#GetEventType() EventType` method, and add a `EventType#Enabled() bool`. So you'd lose the extra stray boolean, and the state would be self-contained in the event type.

Reply:
@yusefnapora take a look at the refactor here: https://github.com/filecoin-project/lotus/pull/2455/commits/bef7d64e666242c19e4f76cb88422c64947e8f00